# Install and import pandas

In [ ]:
!pip install pandas
# Requirement already satisfied: package already installed

In order to use any library in Python, you must first import that library.

In [ ]:
import pandas as pd
# as pd is the shortname

In [ ]:
pd.__version__
# check version

'2.2.2'

We will import student file using read_excel or read_csv functions.

In [ ]:
# import student excel file
pd.read_excel("student.xlsx")

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
# import student csv file
pd.read_csv("student.csv")

In [ ]:
import sys
sys.getsizeof(df)

6363

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_excel("student.xlsx")
# df is short for dataframe.
# this code will save student file in df

In [ ]:
df

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


# Inspecting and Exploring Dataframe

In [ ]:
# top 5 rows
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# top 10 rows
df.head(10)

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
df.head(3)

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male


In [ ]:
# last 5 rows
df.tail()

,id,name,class,mark,gender
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
df.tail(7)

,id,name,class,mark,gender
28,29,Tess Played,Seven,55,male
29,30,Reppy Red,Six,79,female
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
# random sample data
df.sample()

,id,name,class,mark,gender
20,21,Babby John,Four,69,female


In [ ]:
# random sample data
df.sample()

,id,name,class,mark,gender
3,4,Krish Star,Four,60,female


In [ ]:
df.sample(5)

,id,name,class,mark,gender
7,8,Asruid,Five,85,male
19,20,Jackly,Nine,65,female
24,25,Giff Tow,Seven,88,male
28,29,Tess Played,Seven,55,male
14,15,Tade Row,NaN,88,male


In [ ]:
df.sample?
# help on sample method
# open docstring

In [ ]:
df.head?

In [ ]:
# number of rows and columns
df.shape
# 35 rows and 5 column
# .shape doesnt take (). its an attribute
# .sample() is a method

(35, 5)

### Method vs. Attribute in Pandas

In object-oriented programming (which pandas is built upon), objects (like your DataFrame `df`) have both **attributes** and **methods**.

*   **Attributes**: These are characteristics or properties of an object. They store data or provide information about the object's state. You access them directly without parentheses.
    *   **Example**: `df.shape` is an attribute that returns a tuple representing the dimensions (rows, columns) of the DataFrame. `df.index` or `df.columns` are also attributes.

*   **Methods**: These are functions that belong to an object. They perform actions, operations, or calculations on the object's data. You call them using parentheses `()` (even if they don't take any arguments).
    *   **Example**: `df.head()` is a method that returns the first `n` rows of the DataFrame. `df.sample()` and `df.info()` are also methods.

In [ ]:
# dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB


Compare non-null count vs number of entries. if non-null count is different then number of entries, there is a missing value on that column.
- name has 1 missing value
- class has 1 missing value
- gender has 2 missing values

Check datatype for each column. Object is string datatype.

In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# name of the columns
df.columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
# statistical summary
df.describe()

,id,mark
count,35.000000,35.000000
mean,18.000000,74.657143
std,10.246951,16.401117
min,1.000000,18.000000
25%,9.500000,62.500000
50%,18.000000,79.000000
75%,26.500000,88.000000
max,35.000000,96.000000


- mean should be bigger then std (standard deviation).
- 50% is median. In the mark 50% of all student have their mark => 79. 50% of all student have <79.
- 25% of all students have their mark <62.5

In [ ]:
# number of rows
len(df)

35

In [ ]:
# rows * columns
df.size
# 35 * 5

175

In [ ]:
# datatypes of the columns
df.dtypes

,0
id,int64
name,object
class,object
mark,int64
gender,object


In [ ]:
df.isnull()

,id,name,class,mark,gender
0,False,False,False,False,False
1,False,False,False,False,False
2,False,False,False,False,False
3,False,False,False,False,False
4,False,False,False,False,False
5,False,False,False,False,False
6,False,False,False,False,False
7,False,False,False,False,False
8,False,False,False,False,True
9,False,False,False,False,False


In [ ]:
# Counts missing values per column.
df.isnull().sum()

,0
id,0
name,1
class,1
mark,0
gender,2


In [ ]:
# total missing values in df.
df.isnull().sum().sum()

np.int64(4)

# Subsetting DataFrame
- accessing columns and rows

use [] (slicing method) to  access and return the columns

In [ ]:
# return only name column and all rows
df["name"]

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df["class"]

,class
0,Four
1,Three
2,Three
3,Four
4,Four
5,Four
6,Fifth
7,Five
8,Six
9,Four


you can use . notation to return columns.

In [ ]:
df.name

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
df.class
# this will fail. because class is a reserved keyword in python.

SyntaxError: invalid syntax (1584895042.py, line 1)

In [ ]:
df["gender"]

,gender
0,female
1,male
2,male
3,female
4,female
5,male
6,male
7,male
8,NaN
9,female


In [ ]:
df.gender

,gender
0,female
1,male
2,male
3,female
4,female
5,male
6,male
7,male
8,NaN
9,female


if you want to return multiples. you must use []. provide the list of the column name

list objects are created using []

In [ ]:
# name and class
df[["name", "class"]]
# inner [] is list of the column names. ["name", "class"]
# outer [] is for slicing.

,name,class
0,John Deo,Four
1,Max Ruin,Three
2,Arnold,Three
3,Krish Star,Four
4,John Mike,Four
5,Alex John,Four
6,My John Rob,Fifth
7,Asruid,Five
8,Tes Qry,Six
9,Big John,Four


In [ ]:
df["name"].head(3)

,name
0,John Deo
1,Max Ruin
2,Arnold


In [ ]:
df["class"].sample(3)

,class
23,Seven
26,Three
22,Eight


In [ ]:
df["mark"].median()

79.0

In [ ]:
df["mark"].min()

18

In [ ]:
df["mark"].max()

96

In [ ]:
df["mark"].mean()

np.float64(74.65714285714286)

In [ ]:
df[["name", "class"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   name    34 non-null     object
 1   class   34 non-null     object
dtypes: object(2)
memory usage: 692.0+ bytes


In [ ]:
df[["name", "class"]].sample()

,name,class
22,Herod,Eight


# loc and iloc operator

In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
df.iloc[22]

,22
id,23
name,Herod
class,Eight
mark,79
gender,male


In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


df.loc[rows, column labels]

In [ ]:
df.loc[0, "name"]

'John Deo'

In [ ]:
df.loc[3, "mark"]

np.int64(60)

In [ ]:
# names of first 5 students
df.loc[0:4, "name"]

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike


In [ ]:
df.loc[11:14, "name"]

,name
11,Recky
12,Kty
13,Bigy
14,Tade Row


In [ ]:
df.loc[11:14, ["name", "class"]]

,name,class
11,Recky,Six
12,Kty,Seven
13,Bigy,Seven
14,Tade Row,NaN


In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# name of the first
# df.loc[0, "name"]
df.iloc[0, 1]

'John Deo'

In [ ]:
# df.loc[3, "mark"]
df.iloc[3, 3]

np.int64(60)

In [ ]:
# df.loc[11:14, ["name", "class"]]
df.iloc[11:15, [1,2]]

,name,class
11,Recky,Six
12,Kty,Seven
13,Bigy,Seven
14,Tade Row,NaN


- loc operator is inclusive of stopping index.
- iloc operator will exclude the stopping index in the output.

In [ ]:
df.iloc[11:15, 0:2]
# selection of row and column index
# this code will return row indexes 11,12,13 and 14. and also column indexes 0 and 1.

,id,name
11,12,Recky
12,13,Kty
13,14,Bigy
14,15,Tade Row


# Filtering Dataframe

- df[condition]

In [ ]:
# Return student with mark >=90
df[df["mark"]>=90]

,id,name,class,mark,gender
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female


In [ ]:
# Return class Five
df[df["class"]=="Five"]
# first df[] is used for slicing
# second df[] is criteria

,id,name,class,mark,gender
7,8,Asruid,Five,85,male
17,18,Honny,Five,75,male


In [ ]:
# Return class not Five
df[~(df["class"]=="Five")]
# first df[] is used for slicing
# second df[] is criteria

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
# Return class not Five
df[df["class"]!="Five"]
# first df[] is used for slicing
# second df[] is criteria

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
# Return female student at class Four
df[(df["gender"]=="female") & (df["class"]=="Four")]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
20,21,Babby John,Four,69,female


- blue [] is slicing
- brown [] is condition

In [ ]:
# Return female student or student class Four
df[(df["gender"]=="female") | (df["class"]=="Four")]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
15,16,Gimmy,Four,88,male


In [ ]:
# Return student with mark >=90
# df[df["mark"]>=90]
df.loc[df["mark"]>=90, :]
# : means return entire range

,id,name,class,mark,gender
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female


In [ ]:
# df.iloc[df["mark"]>=90, :]

In [ ]:
# Return female student at class Four
# df[(df["gender"]=="female") & (df["class"]=="Four")]
df.loc[(df["gender"]=="female") & (df["class"]=="Four"), :]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
20,21,Babby John,Four,69,female


In [ ]:
df.filter(items=['name', 'mark'])

,name,mark
0,John Deo,75
1,Max Ruin,85
2,Arnold,55
3,Krish Star,60
4,John Mike,60
5,Alex John,55
6,My John Rob,78
7,Asruid,85
8,Tes Qry,78
9,Big John,55


In [ ]:
df.filter(like='ma')
#filter colum names

,mark
0,75
1,85
2,55
3,60
4,60
5,55
6,78
7,85
8,78
9,55


In [ ]:
df.query("mark > 60")

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
14,15,Tade Row,NaN,88,male


In [ ]:
df.query('gender == "male"')

,id,name,class,mark,gender
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
14,15,Tade Row,NaN,88,male
15,16,Gimmy,Four,88,male
16,17,Tumyu,Six,54,male
17,18,Honny,Five,75,male
18,19,Tinny,Nine,18,male


# Sort values

In [ ]:
# Sorting by names
df.loc[:,:].sort_values(by="name")

,id,name,class,mark,gender
5,6,Alex John,Four,55,male
2,3,Arnold,Three,55,male
7,8,Asruid,Five,85,male
20,21,Babby John,Four,69,female
9,10,Big John,Four,55,female
13,14,Bigy,Seven,88,female
31,32,Binn Rott,Seven,90,female
25,26,Crelea,Seven,79,male
33,34,Gain Toe,Seven,69,male
24,25,Giff Tow,Seven,88,male


In [ ]:
# Sorting by mark
df.loc[:,:].sort_values(by="mark")

,id,name,class,mark,gender
18,19,Tinny,Nine,18,male
16,17,Tumyu,Six,54,male
2,3,Arnold,Three,55,male
5,6,Alex John,Four,55,male
28,29,Tess Played,Seven,55,male
21,22,Reggid,Seven,55,female
9,10,Big John,Four,55,female
4,5,John Mike,Four,60,female
3,4,Krish Star,Four,60,female
19,20,Jackly,Nine,65,female


In [ ]:
# Sorting by mark in desc
df.loc[:,:].sort_values(by="mark", ascending=False).head()

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
30,31,Marry Toeey,Four,88,male


In [ ]:
df[["name", "class"]].sort_values(by="class").head()

,name,class
22,Herod,Eight
6,My John Rob,Fifth
17,Honny,Five
7,Asruid,Five
30,Marry Toeey,Four


In [ ]:
# return student who are female and at class seven
df[(df["gender"]=="female") & (df["class"]=="Seven")].sort_values(by="mark")

,id,name,class,mark,gender
21,22,Reggid,Seven,55,female
27,28,Rojj Base,Seven,86,female
13,14,Bigy,Seven,88,female
12,13,Kty,Seven,88,female
31,32,Binn Rott,Seven,90,female


In [ ]:
# return student who are female and at class seven
df[(df["gender"]=="female") & (df["class"]=="Seven")].sort_values(by="mark", ascending=False)

,id,name,class,mark,gender
31,32,Binn Rott,Seven,90,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
27,28,Rojj Base,Seven,86,female
21,22,Reggid,Seven,55,female


In [ ]:
df.sort_values(by = ["mark", "name"]).head()

,id,name,class,mark,gender
18,19,Tinny,Nine,18,male
16,17,Tumyu,Six,54,male
5,6,Alex John,Four,55,male
2,3,Arnold,Three,55,male
9,10,Big John,Four,55,female


In [ ]:
df.sort_values(by = ["mark", "name"], ascending=[False,True]).head()

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
13,14,Bigy,Seven,88,female


# Value counts

In [ ]:
## How many male and female students are there?
df["gender"].value_counts()

,count
gender,
male,17
female,16


In [ ]:
## Class and number of student
df["class"].value_counts()

,count
class,
Seven,10
Four,8
Six,7
Three,3
Nine,2
Five,2
Fifth,1
Eight,1
